src/twisted/internet/defer.py

```diff
@@ -345,6 +345,18 @@ def __del__(self) -> None:
             log.failure(format, self.failResult, debugInfo=debugInfo)
 
 
+@attr.s(frozen=True, slots=True, auto_attribs=True)
+class _Outcome:
+    _result: object
+
+    @_extraneous
+    def unwrap(self) -> object:
+        result = self._result
+        if isinstance(result, Failure):
+            result.raiseException()
+        return result
+
```

Comment:
Adding all these extra allocations and faffing around when it's only required for old versions of Python seems wasteful. All these bugs are fixed on pythons released in the last 5 years. I realize we may need to support these older versions but adding additional runtime overhead in a *very* hot path in every version of python seems like an odd choice. (Gosh I want speedcenter to put commit statuses on PRs so I don't have to hypothesize about performance like this, but allocating a new instance for every single `await` has gotta be bad.)

Reply:
@glyph there's a lot of frame sniffing and tb_next munging going around here that I don't know how to fix otherwise.

Reply:
it's this code here that confounds me the most https://github.com/twisted/twisted/blob/a4fe705c5b450a946966cd63d19b3ba0f8d6cba8/src/twisted/python/failure.py#L502-L574

Reply:
> there's a lot of frame sniffing and tb_next munging going around here that I don't know how to fix otherwise

In the worst case scenario, just have a `_3_7_plus` version of the implementation and swap in the whole function at import time.

Reply:
the test failure still happens on Python 3.10

Reply:
Ah. Hm. In that case I guess this is unfortunate but inevitable. I thought the cpython bugs were fixed ages ago?

Reply:
here's the error log: https://github.com/twisted/twisted/pull/1624/checks?check_run_id=3131779811#step:9:17950